metagpt/rag/factories/base.py

```diff
@@ -0,0 +1,59 @@
+"""Base Factory."""
+
+from typing import Any, Callable
+
+
+class GenericFactory:
+    """Designed to get objects based on any keys."""
+
+    def __init__(self, creators: dict[Any, Callable] = None):
+        """Creators is a dictionary.
+
+        Keys are identifiers, and the values are the associated creator function, which create objects.
+        """
+        self._creators = creators or {}
+
+    def get_instances(self, keys: list[Any], **kwargs) -> list[Any]:
+        """Get instances by keys."""
+        return [self.get_instance(key, **kwargs) for key in keys]
+
+    def get_instance(self, key: Any, **kwargs) -> Any:
+        """Get instance by key.
+
+        Raise Exception if key not found.
+        """
+        creator = self._creators.get(key)
+        if creator:
+            return creator(**kwargs)
+
+        raise ValueError(f"Creator not registered for key: {key}")
+
+
+class ConfigFactory(GenericFactory):
```

Comment:
I'm not sure if this is the common naming, but this name made me thought it was a collection of configs, instead, it's a collection of objects, config is just the key